### PR TITLE
Remove feature: `crate` visibility modifier

### PR DESCRIPTION
**Pull request by jhpratt**
FCP completed in #53120.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/mod.rs [hunk 1]
```diff
@@ -1259,15 +1255,6 @@ impl<'a> Parser<'a> {
         maybe_whole!(self, NtVis, |x| x.into_inner());
 
         self.expected_tokens.push(TokenType::Keyword(kw::Crate));
```
Comment (jhpratt) [inline compiler/rustc_parse/src/parser/mod.rs:1257]: I'm not sure on whether this line should be removed.

File: compiler/rustc_ast/src/ast.rs [hunk 2]
```diff
@@ -2585,7 +2576,7 @@ pub struct Visibility {
 #[derive(Clone, Encodable, Decodable, Debug)]
 pub enum VisibilityKind {
     Public,
-    Crate(CrateSugar),
+    Crate,
     Restricted { path: P<Path>, id: NodeId },
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/ast.rs:2580]: `Crate` can be merged into `Restricted`, it's a special case of it.

(Also span of the `crate` will be properly kept in that case, which can make difference with macro 2.0 hygiene.)

File: compiler/rustc_parse/src/parser/mod.rs (hunk 1 above)
Comment (jhpratt) [reply compiler/rustc_parse/src/parser/mod.rs:1257]: UI tests and `rustc_parse` tests both pass with this removed, so I don't think it's necessary. Removing.

**Comment by jhpratt** [comment]
CI is green.

File: compiler/rustc_ast/src/ast.rs (hunk 2 above)
Comment (jhpratt) [reply compiler/rustc_ast/src/ast.rs:2579]: Done.

File: compiler/rustc_parse/src/parser/mod.rs [hunk 3]
```diff
@@ -1245,30 +1245,15 @@ impl<'a> Parser<'a> {
...
     }
 
-    fn is_crate_vis(&self) -> bool {
-        self.token.is_keyword(kw::Crate) && self.look_ahead(1, |t| t != &token::ModSep)
-    }
-
-    /// Parses `pub`, `pub(crate)` and `pub(in path)` plus shortcuts `crate` for `pub(crate)`,
-    /// `pub(self)` for `pub(in self)` and `pub(super)` for `pub(in super)`.
+    /// Parses `pub`, `pub(crate)` and `pub(in path)` plus shortcuts `pub(self)` for `pub(in self)`
+    /// and `pub(super)` for `pub(in super)`.
```
Comment (fbstj) [inline compiler/rustc_parse/src/parser/mod.rs:1249]:
```suggestion
    /// Parses `pub` and `pub(in path)` plus shortcuts `pub(crate)` for `pub(in crate)`,
    /// `pub(self)` for `pub(in self)` and `pub(super)` for `pub(in super)`.
```

Comment (jhpratt) [reply compiler/rustc_parse/src/parser/mod.rs]: Done.

**Comment by jackh726** [comment]
@bors retry aarch64-gnu seems stuck

It's been running for over 2.5 hours when normally it takes <50 mins.

**Comment by mbartlett21** [comment]
@jhpratt Just a note: The feature doesn't end up even saying it is removed if you have used the `crate` visibility modifier anywhere, since it doesn't parse any more.

For example, in a crate I had that used it, all it gave was this, and no other errors:

```rs
error: expected one of `!` or `::`, found keyword `struct`
  --> src\main.rs:68:11
   |
68 |     crate struct Features(u8);
   |           ^^^^^^ expected one of `!` or `::`
```

**Comment by jhpratt** [comment]
Eh, I don't see an issue with that to be honest. That's what happens when things are removed.

**Pull request merged**